

## Introduction to Data Processing

Welcome to the Data Processing notebook for the Employee Performance Analysis project at INX Future Inc. This notebook is a foundational element of our data analysis pipeline, designed to prepare the raw employee data for in-depth analysis and modeling.

### Objectives:
- **Clean the Data**: Identify and address any issues in the dataset, such as missing values or incorrect data entries.
- **Transform the Data**: Standardize and normalize numerical features and encode categorical variables to ensure that the data is in a suitable format for machine learning algorithms.
- **Optimize Data for Analysis**: Implement transformations that enhance the data's utility for exploring patterns and building predictive models, ensuring that the data is robust and analysis-ready.

In this notebook, we will use a combination of Python libraries, including Pandas and Scikit-learn, to perform these tasks. Our goal is to create a clean, well-structured dataset that will facilitate accurate and insightful analysis in subsequent phases of the project.

Let’s ensure our data is processed effectively to lay a strong foundation for our analysis and modeling efforts.



In [16]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from scipy.sparse import csr_matrix

# Load the dataset
file_path = r'C:\Users\moses_y\OneDrive\Desktop\ML Projects\Employee_Performance_Analysis\data\raw\INX_Future_Inc_Employee_Performance_CDS_Project2_Data_V1.8.xlsx'
data = pd.read_excel(file_path)

# Setting EmpNumber as the index
data.set_index('EmpNumber', inplace=True)

# Define numerical and categorical columns
numerical_cols = data.select_dtypes(include=['int64', 'float64']).columns
categorical_cols = data.select_dtypes(include=['object']).columns

# Define the Pipeline steps for numerical and categorical data preprocessing
numerical_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='median')),
    ('scaler', StandardScaler())
])

categorical_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='most_frequent')),
    ('onehot', OneHotEncoder(handle_unknown='ignore'))
])

# Combine transformers into a ColumnTransformer
preprocessor = ColumnTransformer(
    transformers=[
        ('num', numerical_transformer, numerical_cols),
        ('cat', categorical_transformer, categorical_cols)
    ]
)

# Apply the ColumnTransformer to the dataset
data_preprocessed = preprocessor.fit_transform(data)

# Convert the preprocessed data back to a DataFrame
# If the output is a sparse matrix, convert it to a dense format
if isinstance(data_preprocessed, csr_matrix):
    data_preprocessed = data_preprocessed.toarray()
# Get feature names for transformed categorical columns
categorical_features = preprocessor.named_transformers_['cat']['onehot'].get_feature_names_out(categorical_cols)
all_features = np.append(numerical_cols, categorical_features)

data_preprocessed_df = pd.DataFrame(data_preprocessed, columns=all_features)

# Save the preprocessed data to the processed folder
processed_file_path = r'C:\Users\moses_y\OneDrive\Desktop\ML Projects\Employee_Performance_Analysis\data\processed\INX_Future_Inc_Employee_Performance_Preprocessed.xlsx'
data_preprocessed_df.to_excel(processed_file_path, index=False)

print(f"Preprocessed data has been successfully saved to {processed_file_path}.")


Preprocessed data has been successfully saved to C:\Users\moses_y\OneDrive\Desktop\ML Projects\Employee_Performance_Analysis\data\processed\INX_Future_Inc_Employee_Performance_Preprocessed.xlsx.


In [22]:


# Load the raw data
file_path = r'C:\Users\moses_y\OneDrive\Desktop\ML Projects\Employee_Performance_Analysis\data\raw\INX_Future_Inc_Employee_Performance_CDS_Project2_Data_V1.8.xlsx'
raw_data = pd.read_excel(file_path)
# Setting EmpNumber as the index
raw_data.set_index('EmpNumber', inplace=True)

# Load the processed data
processed_file_path = r'C:\Users\moses_y\OneDrive\Desktop\ML Projects\Employee_Performance_Analysis\data\processed\INX_Future_Inc_Employee_Performance_Preprocessed.xlsx'
processed_data = pd.read_excel(processed_file_path)

# Inspect the structure of both datasets
raw_shape = raw_data.shape
processed_shape = processed_data.shape

# Compare the first few rows of both datasets
raw_head = raw_data.head()
processed_head = processed_data.head()

raw_head


,Age,Gender,EducationBackground,MaritalStatus,EmpDepartment,EmpJobRole,BusinessTravelFrequency,DistanceFromHome,EmpEducationLevel,EmpEnvironmentSatisfaction,...,EmpRelationshipSatisfaction,TotalWorkExperienceInYears,TrainingTimesLastYear,EmpWorkLifeBalance,ExperienceYearsAtThisCompany,ExperienceYearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager,Attrition,PerformanceRating
EmpNumber,,,,,,,,,,,,,,,,,,,,,
E1001000,32,Male,Marketing,Single,Sales,Sales Executive,Travel_Rarely,10,3,4,...,4,10,2,2,10,7,0,8,No,3
E1001006,47,Male,Marketing,Single,Sales,Sales Executive,Travel_Rarely,14,4,4,...,4,20,2,3,7,7,1,7,No,3
E1001007,40,Male,Life Sciences,Married,Sales,Sales Executive,Travel_Frequently,5,4,4,...,3,20,2,3,18,13,1,12,No,4
E1001009,41,Male,Human Resources,Divorced,Human Resources,Manager,Travel_Rarely,10,4,2,...,2,23,2,2,21,6,12,6,No,3
E1001010,60,Male,Marketing,Single,Sales,Sales Executive,Travel_Rarely,16,4,1,...,4,10,1,3,2,2,2,2,No,3


In [23]:
processed_head

,Age,DistanceFromHome,EmpEducationLevel,EmpEnvironmentSatisfaction,EmpHourlyRate,EmpJobInvolvement,EmpJobLevel,EmpJobSatisfaction,NumCompaniesWorked,EmpLastSalaryHikePercent,...,EmpJobRole_Senior Manager R&D,EmpJobRole_Technical Architect,EmpJobRole_Technical Lead,BusinessTravelFrequency_Non-Travel,BusinessTravelFrequency_Travel_Frequently,BusinessTravelFrequency_Travel_Rarely,OverTime_No,OverTime_Yes,Attrition_No,Attrition_Yes
0,-0.541458,0.102061,0.103000,1.177978,-0.543569,0.379608,-0.060955,1.151824,-0.674538,-0.889111,...,0,0,0,0,0,1,1,0,1,0
1,1.109888,0.591464,1.061145,1.177978,-1.187042,0.379608,-0.060955,-1.574386,-0.269410,-0.889111,...,0,0,0,0,0,1,1,0,1,0
2,0.339260,-0.509693,1.061145,1.177978,-0.890055,-1.035081,0.842082,-1.574386,0.945974,1.594054,...,0,0,0,0,1,0,0,1,1,0
3,0.449349,0.102061,1.061145,-0.656641,0.347393,-1.035081,2.648157,1.151824,0.135718,-0.061389,...,0,0,0,0,0,1,1,0,1,0
4,2.541054,0.836165,1.061145,-1.573950,0.891870,0.379608,-0.060955,-1.574386,2.161359,-0.337297,...,0,0,0,0,0,1,1,0,1,0


In [24]:
raw_shape

(1200, 27)

In [25]:
processed_shape

(1200, 62)

The analysis reveals significant differences between the raw and processed datasets:

- **Dataset Shapes:**
  - The raw dataset contains 1200 rows and 27 columns.
  - The processed dataset also consists of 1200 rows but has been expanded to 62 columns. This substantial increase in columns indicates extensive feature engineering,  due to one-hot encoding of categorical variables.

- **Content Overview:**
  - The **raw data** includes a variety of columns such as 'EmpNumber', 'Age', 'Gender', 'EducationBackground', 'MaritalStatus', 'EmpDepartment', 'EmpJobRole', 'BusinessTravelFrequency', among others. These represent a mix of numerical and categorical data.
  - The **processed data** shows a transformation where numerical columns have been scaled (e.g., 'Age' now appears as a standardized value), and categorical columns have been one-hot encoded, resulting in a significant increase in the number of columns. Each unique category in the original dataset's categorical variables now appears as a separate column, with binary indicators (0 or 1) showing the presence or absence of that category for each record.

### Key Observations:

- The preprocessing steps applied have significantly transformed the dataset from its original form to a structure more suitable for machine learning models. This includes normalization/standardization of numerical variables and one-hot encoding of categorical variables, preparing the dataset for algorithms that require numerical input.
- The transformation and expansion of the dataset through one-hot encoding explain the large increase in columns. This is beneficial for models that perform better with distinct features for each category but can also lead to a high-dimensional dataset that may require dimensionality reduction techniques depending on the modeling approach used.
- The presence of scaled numerical data and binary encoded categorical data suggests that the preprocessing has successfully prepared the dataset for further analysis, making it easier to apply various machine learning algorithms without additional data transformation steps.

This thorough preprocessing aligns the dataset with common practices in data science for modeling and analysis, optimizing it for better performance with machine learning algorithms.

,Unnamed: 0,Unnamed: 1
0,NaN,NaN
1,EmpEducationLevel,1 'Below College'
2,NaN,2 'College'
3,NaN,3 'Bachelor'
4,NaN,4 'Master'


(EmpNumber                       0
 Age                             0
 Gender                          0
 EducationBackground             0
 MaritalStatus                   0
 EmpDepartment                   0
 EmpJobRole                      0
 BusinessTravelFrequency         0
 DistanceFromHome                0
 EmpEducationLevel               0
 EmpEnvironmentSatisfaction      0
 EmpHourlyRate                   0
 EmpJobInvolvement               0
 EmpJobLevel                     0
 EmpJobSatisfaction              0
 NumCompaniesWorked              0
 OverTime                        0
 EmpLastSalaryHikePercent        0
 EmpRelationshipSatisfaction     0
 TotalWorkExperienceInYears      0
 TrainingTimesLastYear           0
 EmpWorkLifeBalance              0
 ExperienceYearsAtThisCompany    0
 ExperienceYearsInCurrentRole    0
 YearsSinceLastPromotion         0
 YearsWithCurrManager            0
 Attrition                       0
 PerformanceRating               0
 dtype: int64,
 EmpN

we will implement the following  data analysis steps, we'll proceed through each phase, starting with Data Loading and Cleaning, and progressing through Exploratory Data Analysis (EDA), Feature Selection and Engineering, Model Development, to Insights and Recommendations. 

### 1. Data Loading and Cleaning

The dataset has already been loaded. Now, let's check for missing values, remove any duplicates, and ensure the data types are appropriate.

#### Checking for missing values and duplicates:

In [44]:

# Checking for missing values
missing_values_df2 = df2.isnull().sum()

# Checking for duplicates
duplicates_df2 = df2.duplicated().sum()

missing_values_df2, duplicates_df2

(Unnamed: 0    29
 Unnamed: 1     7
 dtype: int64,
 16)



The dataset contains no missing values or duplicate entries, indicating it's already in a clean state for analysis. This is a great starting point for our exploratory data analysis (EDA).

### Outlier Detection and Correction

In [48]:
## Outlier Detection and Correction with IQR method
# Select numerical columns
numerical_cols = df1.select_dtypes(include=['int64', 'float64']).columns

# Calculate IQR for each column
Q1 = df1[numerical_cols].quantile(0.25)
Q3 = df1[numerical_cols].quantile(0.75)
IQR = Q3 - Q1

# Determine outliers using IQR
outliers = ((df1[numerical_cols] < (Q1 - 1.5 * IQR)) | (df1[numerical_cols] > (Q3 + 1.5 * IQR))).any(axis=1)

# print the proportion of data considered outliers
outlier_proportion = outliers.mean()
outlier_proportion


0.0

In [49]:
## Data Transformation
# we'll standardize the numerical columns using StandardScaler
scaler = StandardScaler()
df1_scaled = df1.copy()
df1_scaled[numerical_cols] = scaler.fit_transform(df1[numerical_cols])

# Checking the scaling effect
scaled_summary = df1_scaled[numerical_cols].describe()
scaled_summary

,Age,DistanceFromHome,EmpEducationLevel,EmpEnvironmentSatisfaction,EmpHourlyRate,EmpJobInvolvement,EmpJobLevel,EmpJobSatisfaction,NumCompaniesWorked,EmpLastSalaryHikePercent,EmpRelationshipSatisfaction,TotalWorkExperienceInYears,TrainingTimesLastYear,EmpWorkLifeBalance,ExperienceYearsAtThisCompany,ExperienceYearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager,PerformanceRating
count,1.200000e+03,1.200000e+03,1.200000e+03,1.200000e+03,1.200000e+03,1.200000e+03,1.200000e+03,1.200000e+03,1.200000e+03,1.200000e+03,1.200000e+03,1.200000e+03,1.200000e+03,1.200000e+03,1.200000e+03,1.200000e+03,1.200000e+03,1.200000e+03,1200.0
mean,-3.523108e-16,-1.628327e-17,-1.051011e-16,-3.256654e-17,-1.213844e-16,2.723747e-16,1.065814e-16,8.881784e-17,-9.473903e-17,-2.960595e-17,-1.243450e-16,1.021405e-16,-2.072416e-16,7.105427e-17,-1.554312e-17,-6.846375e-17,2.368476e-17,-1.036208e-17,0.0
std,1.000417e+00,1.000417e+00,1.000417e+00,1.000417e+00,1.000417e+00,1.000417e+00,1.000417e+00,1.000417e+00,1.000417e+00,1.000417e+00,1.000417e+00,1.000417e+00,1.000417e+00,1.000417e+00,1.000417e+00,1.000417e+00,1.000417e+00,1.000417e+00,0.0
min,-2.082714e+00,-9.990952e-01,-1.813288e+00,-1.573950e+00,-1.781017e+00,-2.449770e+00,-9.639923e-01,-1.574386e+00,-1.090959e+00,-1.165018e+00,-1.604362e+00,-1.520786e+00,-2.150831e+00,-2.494938e+00,-1.261158e+00,-1.205653e+00,-7.789813e-01,-1.172598e+00,0.0
25%,-7.616373e-01,-8.767445e-01,-8.551436e-01,-6.566405e-01,-8.900547e-01,-1.035081e+00,-9.639923e-01,-6.656496e-01,-6.790819e-01,-8.891109e-01,-6.742971e-01,-7.030674e-01,-6.882334e-01,-1.064491e+00,-7.066674e-01,-6.408341e-01,-7.789813e-01,-5.990251e-01,0.0
50%,-1.010990e-01,-2.649912e-01,1.030005e-01,2.606687e-01,9.074614e-04,3.796083e-01,-6.095502e-02,2.430870e-01,-2.672051e-01,-3.372965e-01,2.557678e-01,-1.579219e-01,2.868316e-01,3.659560e-01,-3.370070e-01,-3.584247e-01,-3.744908e-01,-3.122387e-01,0.0
75%,6.695289e-01,5.914635e-01,1.061145e+00,1.177978e+00,8.423717e-01,3.796083e-01,8.420823e-01,1.151824e+00,5.565486e-01,7.663322e-01,1.185833e+00,5.235101e-01,2.868316e-01,3.659560e-01,5.871439e-01,7.712132e-01,4.344902e-01,8.349069e-01,0.0
max,2.541054e+00,2.426724e+00,2.019289e+00,1.177978e+00,1.683836e+00,1.794297e+00,2.648157e+00,1.151824e+00,2.409994e+00,2.697683e+00,1.185833e+00,2.363376e+00,1.749429e+00,1.796403e+00,2.527861e+00,2.889284e+00,2.254698e+00,2.985805e+00,0.0


#### Data Cleaning and Processing:

Let's begin with Data Cleaning and Preprocessing to ensure our dataset is ready for analysis. This step will include checking for missing values, encoding categorical variables, and addressing any anomalies present in the dataset. Let's start by encoding the categorical variables to prepare our dataset for further analysis.



In [5]:

from sklearn.preprocessing import LabelEncoder

# Initialize LabelEncoder
le = LabelEncoder()

# Identifying categorical columns (excluding 'EmpNumber' as it's a unique identifier)
categorical_cols = df1.select_dtypes(include=['object']).columns.drop('EmpNumber')

# Apply LabelEncoder to each categorical column
for col in categorical_cols:
    df1[col] = le.fit_transform(df1[col])

# Display the first few rows to verify the encoding
df1.head()



### 2. Exploratory Data Analysis (EDA)

#### Summary Statistics:


,Age,DistanceFromHome,EmpEducationLevel,EmpEnvironmentSatisfaction,EmpHourlyRate,EmpJobInvolvement,EmpJobLevel,EmpJobSatisfaction,NumCompaniesWorked,EmpLastSalaryHikePercent,EmpRelationshipSatisfaction,TotalWorkExperienceInYears,TrainingTimesLastYear,EmpWorkLifeBalance,ExperienceYearsAtThisCompany,ExperienceYearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager,PerformanceRating
count,1200.000000,1200.000000,1200.00000,1200.000000,1200.000000,1200.000000,1200.000000,1200.000000,1200.000000,1200.000000,1200.000000,1200.000000,1200.000000,1200.000000,1200.000000,1200.000000,1200.000000,1200.000000,1200.000000
mean,36.918333,9.165833,2.89250,2.715833,65.981667,2.731667,2.067500,2.732500,2.665000,15.222500,2.725000,11.330000,2.785833,2.744167,7.077500,4.291667,2.194167,4.105000,2.948333
std,9.087289,8.176636,1.04412,1.090599,20.211302,0.707164,1.107836,1.100888,2.469384,3.625918,1.075642,7.797228,1.263446,0.699374,6.236899,3.613744,3.221560,3.541576,0.518866
min,18.000000,1.000000,1.00000,1.000000,30.000000,1.000000,1.000000,1.000000,0.000000,11.000000,1.000000,0.000000,0.000000,1.000000,0.000000,0.000000,0.000000,0.000000,2.000000
25%,30.000000,2.000000,2.00000,2.000000,48.000000,2.000000,1.000000,2.000000,1.000000,12.000000,2.000000,6.000000,2.000000,2.000000,3.000000,2.000000,0.000000,2.000000,3.000000
50%,36.000000,7.000000,3.00000,3.000000,66.000000,3.000000,2.000000,3.000000,2.000000,14.000000,3.000000,10.000000,3.000000,3.000000,5.000000,3.000000,1.000000,3.000000,3.000000
75%,43.000000,14.000000,4.00000,4.000000,83.000000,3.000000,3.000000,4.000000,4.000000,18.000000,4.000000,15.000000,3.000000,3.000000,10.000000,7.000000,3.000000,7.000000,3.000000
max,60.000000,29.000000,5.00000,4.000000,100.000000,4.000000,5.000000,4.000000,9.000000,25.000000,4.000000,40.000000,6.000000,4.000000,40.000000,18.000000,15.000000,17.000000,4.000000



The summary statistics provide insights into the numerical features of the dataset, including measures of central tendency and variability. Here are some key takeaways:

- **Age**: The average age of employees is approximately 37 years, with a spread from 18 to 60 years.
- **DistanceFromHome**: On average, employees live about 9.17 units of distance from their workplace. This varies widely, with a maximum distance of 29 units.
- **EmpEducationLevel** & **EmpEnvironmentSatisfaction**: These features, along with several others like EmpJobLevel and EmpJobSatisfaction, are measured on scales (likely 1 to 5 or 1 to 4), indicating levels of education, job satisfaction, and environmental satisfaction.
- **NumCompaniesWorked**: Employees have worked with up to 9 different companies, but on average, have experience with about 2.67 companies.
- **TotalWorkExperienceInYears**: On average, employees have 11.33 years of total work experience, ranging from 0 to 40 years.
- **PerformanceRating**: The performance ratings are likely on a scale, with an average rating of approximately 2.95, indicating that most employees fall into a middle category if the scale is from 1 to 4.

#### Correlation Analysis:

ValueError: could not convert string to float: 'E1001000'

The correlation analysis reveals how various features relate to the `PerformanceRating`. Here are some notable observations:

- **Positive Correlations**: Certain features show a positive correlation with `PerformanceRating`, meaning that as these features increase, so does the performance rating.
  - `EmpEnvironmentSatisfaction` and `EmpLastSalaryHikePercent` are the most positively correlated with performance rating, indicating that employees who are more satisfied with their work environment and those who have received significant salary hikes tend to have higher performance ratings.
  - `EmpWorkLifeBalance` also shows a positive correlation, suggesting that employees who feel they have a good balance between work and personal life tend to perform better.

- **Negative Correlations**: Several features are negatively correlated with performance rating, indicating that higher values in these features could relate to lower performance ratings.
  - `YearsSinceLastPromotion`, `ExperienceYearsInCurrentRole`, `YearsWithCurrManager`, and `ExperienceYearsAtThisCompany` are negatively correlated. This might suggest that employees who have been in the same role or with the same manager for longer periods, or those who have not been promoted recently, tend to have lower performance ratings.
  - Surprisingly, `EmpJobLevel` and `TotalWorkExperienceInYears` also have a slight negative correlation with performance rating, which could indicate that higher job levels or more work experience do not necessarily translate to higher performance ratings in this dataset.

Given these insights, we can now move on to the feature selection and engineering phase. Based on the correlation analysis, `EmpEnvironmentSatisfaction`, `EmpLastSalaryHikePercent`, and `EmpWorkLifeBalance` emerge as significant predictors for employee performance. Additionally, considering creating new features or transforming existing ones to better capture their relationship with employee performance could be beneficial.

#### Visualizations:

visualizations that would be valuable:

- **Histograms**: Useful for understanding the distribution of numerical variables like Age, DistanceFromHome, etc.
- **Scatter Plots**: Can help identify relationships between continuous variables and PerformanceRating.
- **Box Plots**: Effective for detecting outliers in the data, for instance, in TotalWorkExperienceInYears.

## Feature Engineering

Feature engineering is a crucial step in data analysis and model development, as it can significantly improve model performance by creating new features that capture important information in more effective ways.

Given the insights from our correlation analysis, let's consider a few strategies for feature engineering:

1. **Interaction Terms**: Since `EmpEnvironmentSatisfaction`, `EmpLastSalaryHikePercent`, and `EmpWorkLifeBalance` are strongly correlated with performance, creating interaction terms between these variables might highlight employees who are particularly likely to have high or low performance ratings. For example, an interaction term between `EmpEnvironmentSatisfaction` and `EmpLastSalaryHikePercent` could capture the combined effect of these variables on performance.

2. **Aggregated Metrics**: For employees with longer tenure at the company (`ExperienceYearsAtThisCompany`), considering their average performance rating over the years or changes in performance rating over time might provide insights into their growth or consistency in performance.

3. **Categorical Bins**: Converting continuous variables into categorical bins can sometimes improve model performance by simplifying the relationships. For instance, `Age` could be binned into categories such as 'Young', 'Mid-age', and 'Senior' to see if these groups differ significantly in performance.

Let's implement feature engineering: creating interaction terms for `EmpEnvironmentSatisfaction` and `EmpLastSalaryHikePercent`, and then we'll prepare the dataset for model development.

In [12]:

# Creating an interaction term between EmpEnvironmentSatisfaction and EmpLastSalaryHikePercent
df['Satisfaction_Hike_Interaction'] = df['EmpEnvironmentSatisfaction'] * df['EmpLastSalaryHikePercent']

# Additionally, let's bin the 'Age' as an example of creating categorical bins
df['AgeGroup'] = pd.cut(df['Age'], bins=[0, 30, 50, 100], labels=['Young', 'Mid-age', 'Senior'])

# Converting 'AgeGroup' to a categorical type
df['AgeGroup'] = df['AgeGroup'].astype('category')

# Displaying the first few rows to see the new features
df[['EmpEnvironmentSatisfaction', 'EmpLastSalaryHikePercent', 'Satisfaction_Hike_Interaction', 'Age', 'AgeGroup']].head()


,EmpEnvironmentSatisfaction,EmpLastSalaryHikePercent,Satisfaction_Hike_Interaction,Age,AgeGroup
0,4,12,48,32,Mid-age
1,4,12,48,47,Mid-age
2,4,21,84,40,Mid-age
3,2,15,30,41,Mid-age
4,1,14,14,60,Senior


This step adds valuable new features to our dataset. Now, we would typically proceed to model development, where these features can be included to evaluate their impact on predicting employee performance.

In [15]:
# Creating an interaction term between EmpEnvironmentSatisfaction and EmpLastSalaryHikePercent
df['Satisfaction_Hike_Interaction'] = df['EmpEnvironmentSatisfaction'] * df['EmpLastSalaryHikePercent']

# Additionally, binning 'Age' and converting 'AgeGroup' to a categorical type
df['AgeGroup'] = pd.cut(df['Age'], bins=[0, 30, 50, 100], labels=['Young', 'Mid-age', 'Senior'])
df['AgeGroup'] = df['AgeGroup'].astype('category')

# Performing correlation analysis with the new feature
correlation_matrix_new = df.corr()

# Plotting the heatmap of the correlation matrix
plt.figure(figsize=(12, 8))
sns.heatmap(correlation_matrix_new, annot=True, fmt=".2f", cmap="coolwarm")
plt.title("Correlation Matrix with New Features")
plt.show()

# Focusing on the correlation with PerformanceRating, including the new feature
performance_correlation_new = correlation_matrix_new['PerformanceRating'].sort_values()
performance_correlation_new


ValueError: could not convert string to float: 'E1001000'

The correlation analysis, now including our newly engineered feature `Satisfaction_Hike_Interaction`, reveals some interesting insights:

- The `Satisfaction_Hike_Interaction` feature, which combines `EmpEnvironmentSatisfaction` and `EmpLastSalaryHikePercent`, has a strong positive correlation of approximately 0.53 with `PerformanceRating`. This suggests that the interaction between an employee's satisfaction with their environment and their last salary hike is a significant predictor of their performance rating, even more so than either component individually. This aligns with our hypothesis that combining these factors could capture a more nuanced effect on performance.

- As expected, the original features `EmpEnvironmentSatisfaction` and `EmpLastSalaryHikePercent` remain positively correlated with performance, but the interaction term's correlation is notably higher, underscoring the value of feature engineering in uncovering deeper insights.

- The heatmap visualized with Seaborn provides a comprehensive overview of how all features correlate with each other, with our focus being on the `PerformanceRating`. The new feature stands out for its strong relationship with performance, suggesting that it would be a valuable inclusion in any model aiming to predict employee performance.

Given these findings, the engineered feature could play a critical role in model development, potentially improving the accuracy of predictions for employee performance ratings.

With these insights in hand, we can proceed with model development, keeping in mind the significance of the `Satisfaction_Hike_Interaction` feature. 


Given the findings from our exploratory data analysis (EDA) and feature engineering, we can derive several actionable insights and recommendations for INX Future Inc. to enhance employee performance. Here are some key insights and the corresponding strategic recommendations:

### Insights and Recommendations

1. **Environment Satisfaction and Salary Hike Interaction**:
   - **Insight**: The interaction between employee environment satisfaction and the last salary hike is strongly correlated with performance ratings.
   - **Recommendation**: Prioritize initiatives that enhance workplace satisfaction while aligning them with fair and transparent compensation policies. Consider implementing personalized reward systems that recognize both individual achievements and contributions to a positive work environment.

2. **Importance of Work-Life Balance**:
   - **Insight**: Work-life balance has a positive correlation with employee performance ratings.
   - **Recommendation**: Develop programs and policies that support flexible working hours, remote work options, and mental health days. Promoting a culture that values personal time and life outside work can lead to higher productivity and satisfaction.

3. **Impact of Long Tenure Without Promotion**:
   - **Insight**: Longer tenure without promotion, stagnant roles, and prolonged periods with the same manager negatively correlate with performance ratings.
   - **Recommendation**: Implement regular career pathing reviews to ensure employees have clear advancement opportunities. Encourage lateral moves to broaden skills and experiences. Training managers to effectively recognize and advocate for their team's growth can also mitigate these negative impacts.

## Model Development Strategies

For developing a predictive model of employee performance, considering both classification and regression approaches could be beneficial, depending on the specific goals (e.g., identifying employees at risk of low performance vs. predicting exact performance ratings).

1. **Classification Approach** (e.g., High/Medium/Low Performance):
   - **Models**: Decision Trees, Random Forest, Gradient Boosting Machines (GBM), and Support Vector Machines (SVM).
   - **Rationale**: These models can classify employees into predefined performance categories, useful for targeted interventions.

2. **Regression Approach** (Predicting Performance Score):
   - **Models**: Linear Regression, Ridge/Lasso Regression, and Elastic Net for simpler relationships. For more complex patterns, consider using Random Forest Regressor or Gradient Boosting Regressor.
   - **Rationale**: If performance ratings are numeric and a more precise prediction is required, regression models can estimate the actual performance score.

## Preparing for Model Development

Before model training:
- **Split the Data**: Divide the dataset into training and test sets to evaluate the model's performance on unseen data.
- **Feature Preprocessing**: Apply feature scaling and encoding as necessary, especially for models sensitive to feature magnitude (e.g., SVM) or when using distance-based algorithms.
- **Cross-Validation**: Use cross-validation techniques to ensure the model's robustness and generalize well beyond the training dataset.

### Preprocessing

Since we already performed some feature engineering, we'll proceed with splitting the data and encoding categorical variables as necessary.

Let's begin with the **Classification Approach**. For this example, we'll convert the `PerformanceRating` into three categories (assuming 1-2: Low, 3: Medium, 4: High) for classification purposes. Then, we'll encode categorical variables if needed and split the data. 






Given the objective to build two models—one for classification and another for regression—we'll outline the steps and provide example code for each approach. We'll use Random Forest for classification due to its robustness and ability to handle unbalanced data, and Gradient Boosting Regressor for regression, known for its high performance on a wide range of problems.

### 1. Classification Approach with Random Forest

Objective: Classify employees into performance categories (High/Medium/Low).

**Steps**:
- Preprocess the data: Encode categorical variables, if not already done.
- Split the dataset into training and test sets.
- Train a Random Forest classifier.
- Evaluate the model using accuracy, precision, recall, and F1-score.

### Classification Model Code

In [18]:


# Assuming 'df' is your DataFrame and 'PerformanceRating' is the target variable
# Converting 'PerformanceRating' into a categorical target for classification
df['PerformanceClass'] = pd.cut(df['PerformanceRating'], bins=[0, 2, 3, 4], labels=['Low', 'Medium', 'High'])

# Encoding categorical variables using LabelEncoder or pd.get_dummies
encoder = LabelEncoder()
for col in df.select_dtypes(include=['object', 'category']).columns:
    df[col] = encoder.fit_transform(df[col])

# Splitting the data
X = df.drop(['PerformanceRating', 'PerformanceClass'], axis=1)
y = df['PerformanceClass']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Training the model
classifier = RandomForestClassifier(random_state=42)
classifier.fit(X_train, y_train)

# Predicting and evaluating the model
y_pred = classifier.predict(X_test)
print(classification_report(y_test, y_pred))


              precision    recall  f1-score   support

           0       0.95      0.70      0.81        27
           1       0.93      0.90      0.91        29
           2       0.94      0.98      0.96       184

    accuracy                           0.94       240
   macro avg       0.94      0.86      0.89       240
weighted avg       0.94      0.94      0.94       240




### 2. Regression Approach with Gradient Boosting Regressor

Objective: Predict the exact performance score.

**Steps**:
- Preprocess the data: Apply feature scaling if necessary.
- Split the dataset into training and test sets.
- Train a Gradient Boosting Regressor.
- Evaluate the model using RMSE (Root Mean Square Error) and R² score.

### Regression Model


In [19]:


# Assuming 'df' is your DataFrame and 'PerformanceRating' is the numeric target
X = df.drop(['PerformanceRating', 'PerformanceClass'], axis=1)  # Assuming 'PerformanceClass' was added for classification
y = df['PerformanceRating']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Training the model
regressor = GradientBoostingRegressor(random_state=42)
regressor.fit(X_train, y_train)

# Predicting and evaluating the model
y_pred = regressor.predict(X_test)
print("RMSE:", mean_squared_error(y_test, y_pred, squared=False))
print("R² score:", r2_score(y_test, y_pred))

RMSE: 0.24915368341598068
R² score: 0.7338741188983595
